### PR TITLE
[unit test][CLI][pfcwd] Added pfcwd config tests for single and multi ASIC platform.

### PR DESCRIPTION
**Pull request by smaheshm**
**- What I did**
Added pfcwd config tests for single and multi ASIC platform.

Changed the pfcwd CLI to access the DB object as a parameter.

**- How I did it ** 
Updated DB object that creates DB client connections for Config DBs and other redis DBs to include db/config clients for all namespaces on a multi ASIC platform and for default namespace on a single ASIC platform.

'run_on_multi_asic' decorator uses this object to reuse the DB client connections, thus always reading the mock redis DB in the memory, instead of loading config DB from file. 

**- How to verify it**
Unit tests
```
samaddik@e3cf370f3adb:/sonic/src/sonic-utilities/tests$ sudo pytest -k pfcwd_test.py                                                                                                  
================================================================================ test session starts =================================================================================
platform linux2 -- Python 2.7.16, pytest-3.10.1, py-1.7.0, pluggy-0.8.0                                                                                                               
rootdir: /sonic/src/sonic-utilities/tests, inifile: pytest.ini                                                                                                                        
plugins: cov-2.6.0                                                                                                                                                                    
collected 339 items / 323 deselected                                                                                                                                                  
                                                                                                                                                                                      
pfcwd_test.py ................                                                                                                                                                 [100%] 
                                                                                                                                                                                      
===================================================================== 16 passed, 323 deselected in 1.65 seconds ======================================================================
Exception TypeError: "'NoneType' object is not callable" in <bound method Logger.__del__ of <sonic_py_common.logger.Logger object at 0x7ff9d2baead0>> ignored                         
Exception TypeError: "'NoneType' object is not callable" in <bound method Logger.__del__ of <sonic_py_common.logger.Logger object at 0x7ff9d3832b10>> ignored                         
Exception TypeError: "'NoneType' object is not callable" in <bound method Logger.__del__ of <sonic_py_common.logger.Logger object at 0x7ff9d3ddbc50>> ignored     


Last login: Mon Dec  7 16:09:14 2020 from 100.127.20.23
admin@str-s6000-acs-9:~$ sudo pfcwd start --action drop --restoration-time 199 all 199
admin@str-s6000-acs-9:~$ show pfcwd config
       PORT    ACTION    DETECTION TIME    RESTORATION TIME
-----------  --------  ----------------  ------------------
  Ethernet0      drop               199                 199
  Ethernet4      drop               199                 199
  Ethernet8      drop               199                 199
 Ethernet12      drop               199                 199
 Ethernet16      drop               199                 199
 Ethernet20      drop               199                 199
 Ethernet24      drop               199                 199
 Ethernet28      drop               199                 199
 Ethernet32      drop               199                 199
 Ethernet36      drop               199                 199
 Ethernet40      drop               199                 199
 Ethernet44      drop               199                 199
 Ethernet48      drop               199                 199
 Ethernet52      drop               199                 199
 Ethernet56      drop               199                 199
 Ethernet60      drop               199                 199
 Ethernet64      drop               199                 199
 Ethernet68      drop               199                 199
 Ethernet72      drop               199                 199
 Ethernet76      drop               199                 199
 Ethernet80      drop               199                 199
 Ethernet84      drop               199                 199
 Ethernet88      drop               199                 199
 Ethernet92      drop               199                 199
 Ethernet96      drop               199                 199
Ethernet100      drop               199                 199
Ethernet104      drop               199                 199
Ethernet108      drop               199                 199
Ethernet112      drop               199                 199
Ethernet116      drop               199                 199
Ethernet120      drop               199                 199
Ethernet124      drop               199                 199
admin@str-s6000-acs-9:~$ 


```
**- Previous command output (if the output of a command-line utility has changed)**

**- New command output (if the output of a command-line utility has changed)**

### CODE REVIEW

**Comment by smaheshm** [comment]
👋

**Comment by smaheshm** [comment]
retest this please

**Comment by smaheshm** [comment]
retest this please

**Comment by smaheshm** [comment]
retest this please

**Comment by smaheshm** [comment]
@arlakshm Thanks for providing inputs. Updated as per discussion.

File: pfcwd/main.py [hunk 1]
```diff
@@ -98,10 +100,16 @@ def get_server_facing_ports(db):
...
 class PfcwdCli(object):
-    def __init__(self, namespace=None, display=constants.DISPLAY_ALL):
+    def __init__(
+        self, ctx, db=None, namespace=None, display=constants.DISPLAY_ALL
+    ):
         self.db = None
         self.config_db = None
-        self.multi_asic = multi_asic_util.MultiAsic(display, namespace)
+        # ctx.obj contains the Db object set by clicommon.pass_db decorator
+        # Unit tests over write the ctx.obj with test Db object, using ctx.obj
```
Comment (judyjoseph) [inline pfcwd/main.py:109]: Do we really need to pass both ctx and db ? With the  @clicommon.pass_db decorator, the "db" should have the db object which we created and pass it into runner object in test scripts ?

Comment (smaheshm) [reply pfcwd/main.py:109]: I'll test it out and find out if both are required. I expect "db" will be populated (with single ASIC) when the modules are loaded initially, before the masic env variable is set.

**Comment by smaheshm** [comment]
retest this please

**Review by arlakshm — APPROVED** [review]
lgtm

File: tests/pfcwd_test.py [hunk 2]
```diff
@@ -142,6 +193,67 @@ def test_pfcwd_config_with_ports(self):
...
+        assert result.output == show_pfc_config_start_pass
+
+    def test_pfcwd_start_ports_masic_invalid(self):
+        # --action drop --restoration-time 200 Ethernet0 Ethernet500 200
+        import pfcwd.main as pfcwd
+        runner = CliRunner()
+        db = Db()
+
+        result = runner.invoke(
+            pfcwd.cli.commands["start"],
```
Comment (neethajohn) [inline tests/pfcwd_test.py:236]: Isn't the behavior to fail the cmd if something is incorrect? if that's the case, can you add different action and restoration, detection times for Ethernet0 and still the output should be unchanged

File: tests/pfcwd_test.py [hunk 3]
```diff
@@ -142,6 +193,67 @@ def test_pfcwd_config_with_ports(self):
...
+                "--action", "forward", "--restoration-time", "101",
+                "Ethernet0", "Ethernet-500", "102"
+            ],
+            obj=db
+        )
+        print(result.output)
+        assert result.exit_code == 0
+        assert result.output == show_pfc_config_start_fail
+
+        # get config after the command, config shoudln't change
```
Comment (neethajohn) [inline tests/pfcwd_test.py:247]: spell check "shouldn't"

**Comment by neethajohn** [comment]
Also can you fix the description since you are not creating a new multi asic db object but extending the existing one?

File: tests/pfcwd_test.py (hunk 2 above)
Comment (smaheshm) [reply tests/pfcwd_test.py:236]: click handles the options for actions, so will throw an error if invalid option is specified.

Comment (neethajohn) [reply tests/pfcwd_test.py:236]: would be good to add Ethernet0 in this command with diff action and times

Comment (smaheshm) [reply tests/pfcwd_test.py:383]: done

**Review by neethajohn — APPROVED** [review]

**Review by judyjoseph — APPROVED** [review]

**Comment by smaheshm** [comment]
retest this please

**Comment by smaheshm** [comment]
retest this please

**Comment by stephenxs** [comment]
Hi @smaheshm, 
I'm encountering the following errors well compiling the sonic-utilities.
Looks like it's related to this PR. Could you please check it?

```
=================================== FAILURES ===================================
____________________ TestPfcwd.test_pfcwd_start_ports_valid ____________________

self = <tests.pfcwd_test.TestPfcwd object at 0x7f1222030b00>

    def test_pfcwd_start_ports_valid(self):
        # pfcwd start --action drop --restoration-time 200 Ethernet0 200
        import pfcwd.main as pfcwd
        runner = CliRunner()
        db = Db()
    
        # get initial config
        result = runner.invoke(
            pfcwd.cli.commands["show"].commands["config"],
            obj=db
        )
        print(result.output)
        assert result.output == pfcwd_show_config_output
    
        result = runner.invoke(
            pfcwd.cli.commands["start"],
            [
                "--action", "forward", "--restoration-time", "101",
                "Ethernet0", "102"
            ],
            obj=db
        )
        print(result.output)
>       assert result.exit_code == 0
E       AssertionError: assert 1 == 0
E        +  where 1 = <Result SystemExit('Root privileges are required for this operation')>.exit_code

tests/pfcwd_test.py:104: AssertionError
----------------------------- Captured stdout call -----------------------------
Changed polling interval to 600ms
     PORT    ACTION    DETECTION TIME    RESTORATION TIME
---------  --------  ----------------  ------------------
Ethernet0      drop               600                 600
Ethernet4      drop               600                 600
Ethernet8      drop               600                 600

Root privileges are required for this operation

______________________ TestPfcwd.test_pfcwd_start_actions ______________________

self = <tests.pfcwd_test.TestPfcwd object at 0x7f12229b2048>

    def test_pfcwd_start_actions(self):
        # pfcwd start --action fwd --restoration-time 200 Ethernet0 200
        import pfcwd.main as pfcwd
        runner = CliRunner()
        db = Db()
    
        # get initial config
        result = runner.invoke(
            pfcwd.cli.commands["show"].commands["config"],
            obj=db
        )
        print(result.output)
        assert result.output == pfcwd_show_config_output
    
        result = runner.invoke(
            pfcwd.cli.commands["start"],
            [
                "--action", "forward", "--restoration-time", "301",
                "all", "302"
            ],
            obj=db
        )
        print(result.output)
>       assert result.exit_code == 0
E       AssertionError: assert 1 == 0
E        +  where 1 = <Result SystemExit('Root privileges are required for this operation')>.exit_code

tests/pfcwd_test.py:138: AssertionError
----------------------------- Captured stdout call -----------------------------
Changed polling interval to 600ms
     PORT    ACTION    DETECTION TIME    RESTORATION TIME
---------  --------  ----------------  ------------------
Ethernet0      drop               600                 600
Ethernet4      drop               600                 600
Ethernet8      drop               600                 600

Root privileges are required for this operation

__________ TestMultiAsicPfcwdShow.test_pfcwd_start_ports_masic_valid ___________

self = <tests.pfcwd_test.TestMultiAsicPfcwdShow object at 0x7f12221c2d30>

    def test_pfcwd_start_ports_masic_valid(self):
        # pfcwd start --action forward --restoration-time 200 Ethernet0 200
        import pfcwd.main as pfcwd
        runner = CliRunner()
        db = Db()
        # get initial config
        result = runner.invoke(
            pfcwd.cli.commands["show"].commands["config"],
            obj=db
        )
        print(result.output)
        assert result.output == show_pfc_config_all
    
        result = runner.invoke(
            pfcwd.cli.commands["start"],
            [
                "--action", "forward", "--restoration-time", "101",
                "Ethernet0", "Ethernet-BP4", "102"
            ],
            obj=db
        )
        print(result.output)
>       assert result.exit_code == 0
E       AssertionError: assert 1 == 0
E        +  where 1 = <Result SystemExit('Root privileges are required for this operation')>.exit_code

tests/pfcwd_test.py:292: AssertionError
----------------------------- Captured stdout call -----------------------------
Changed polling interval to 199ms on asic0
BIG_RED_SWITCH status is enable on asic0
Changed polling interval to 199ms on asic1
BIG_RED_SWITCH status is enable on asic1
          PORT    ACTION    DETECTION TIME    RESTORATION TIME
--------------  --------  ----------------  ------------------
     Ethernet0      drop               200                 200
     Ethernet4      drop               200                 200
  Ethernet-BP0      drop               200                 200
  Ethernet-BP4      drop               200                 200
Ethernet-BP256      drop               200                 200
Ethernet-BP260      drop               200                 200

Root privileges are required for this operation

____________ TestMultiAsicPfcwdShow.test_pfcwd_start_actions_masic _____________

self = <tests.pfcwd_test.TestMultiAsicPfcwdShow object at 0x7f1221eb7400>

    def test_pfcwd_start_actions_masic(self):
        # pfcwd start --action drop --restoration-time 200 Ethernet0 200
        import pfcwd.main as pfcwd
        runner = CliRunner()
        db = Db()
        # get initial config
        result = runner.invoke(
            pfcwd.cli.commands["show"].commands["config"],
            obj=db
        )
        print(result.output)
        assert result.output == show_pfc_config_all
    
        result = runner.invoke(
            pfcwd.cli.commands["start"],
            [
                "--action", "drop", "--restoration-time", "301",
                "all", "302"
            ],
            obj=db
        )
        print(result.output)
>       assert result.exit_code == 0
E       AssertionError: assert 1 == 0
E        +  where 1 = <Result SystemExit('Root privileges are required for this operation')>.exit_code

tests/pfcwd_test.py:325: AssertionError
```

**Comment by smaheshm** [comment]
> Hi @smaheshm,
> I'm encountering the following errors well compiling the sonic-utilities.
> Looks like it's related to this PR. Could you please check it?
> 
> ```
> =================================== FAILURES ===================================
> ____________________ TestPfcwd.test_pfcwd_start_ports_valid ____________________
> 
> self = <tests.pfcwd_test.TestPfcwd object at 0x7f1222030b00>
> 
>     def test_pfcwd_start_ports_valid(self):
>         # pfcwd start --action drop --restoration-time 200 Ethernet0 200
>         import pfcwd.main as pfcwd
>         runner = CliRunner()
>         db = Db()
>     
>         # get initial config
>         result = runner.invoke(
>             pfcwd.cli.commands["show"].commands["config"],
>             obj=db
>         )
>         print(result.output)
>         assert result.output == pfcwd_show_config_output
>     
>         result = runner.invoke(
>             pfcwd.cli.commands["start"],
>             [
>                 "--action", "forward", "--restoration-time", "101",
>                 "Ethernet0", "102"
>             ],
>             obj=db
>         )
>         print(result.output)
> >       assert result.exit_code == 0
> E       AssertionError: assert 1 == 0
> E        +  where 1 = <Result SystemExit('Root privileges are required for this operation')>.exit_code
> 
> tests/pfcwd_test.py:104: AssertionError
> ----------------------------- Captured stdout call -----------------------------
> Changed polling interval to 600ms
>      PORT    ACTION    DETECTION TIME    RESTORATION TIME
> ---------  --------  ----------------  ------------------
> Ethernet0      drop               600                 600
> Ethernet4      drop               600                 600
> Ethernet8      drop               600                 600
> 
> Root privileges are required for this operation
> 
> ______________________ TestPfcwd.test_pfcwd_start_actions ______________________
> 
> self = <tests.pfcwd_test.TestPfcwd object at 0x7f12229b2048>
> 
>     def test_pfcwd_start_actions(self):
>         # pfcwd start --action fwd --restoration-time 200 Ethernet0 200
>         import pfcwd.main as pfcwd
>         runner = CliRunner()
>         db = Db()
>     
>         # get initial config
>         result = runner.invoke(
>             pfcwd.cli.commands["show"].commands["config"],
>             obj=db
>         )
>         print(result.output)
>         assert result.output == pfcwd_show_config_output
>     
>         result = runner.invoke(
>             pfcwd.cli.commands["start"],
>             [
>                 "--action", "forward", "--restoration-time", "301",
>                 "all", "302"
>             ],
>             obj=db
>         )
>         print(result.output)
> >       assert result.exit_code == 0
> E       AssertionError: assert 1 == 0
> E        +  where 1 = <Result SystemExit('Root privileges are required for this operation')>.exit_code
> 
> tests/pfcwd_test.py:138: AssertionError
> ----------------------------- Captured stdout call -----------------------------
> Changed polling interval to 600ms
>      PORT    ACTION    DETECTION TIME    RESTORATION TIME
> ---------  --------  ----------------  ------------------
> Ethernet0      drop               600                 600
> Ethernet4      drop               600                 600
> Ethernet8      drop               600                 600
> 
> Root privileges are required for this operation
> 
> __________ TestMultiAsicPfcwdShow.test_pfcwd_start_ports_masic_valid ___________
> 
> self = <tests.pfcwd_test.TestMultiAsicPfcwdShow object at 0x7f12221c2d30>
> 
>     def test_pfcwd_start_ports_masic_valid(self):
>         # pfcwd start --action forward --restoration-time 200 Ethernet0 200
>         import pfcwd.main as pfcwd
>         runner = CliRunner()
>         db = Db()
>         # get initial config
>         result = runner.invoke(
>             pfcwd.cli.commands["show"].commands["config"],
>             obj=db
>         )
>         print(result.output)
>         assert result.output == show_pfc_config_all
>     
>         result = runner.invoke(
>             pfcwd.cli.commands["start"],
>             [
>                 "--action", "forward", "--restoration-time", "101",
>                 "Ethernet0", "Ethernet-BP4", "102"
>             ],
>             obj=db
>         )
>         print(result.output)
> >       assert result.exit_code == 0
> E       AssertionError: assert 1 == 0
> E        +  where 1 = <Result SystemExit('Root privileges are required for this operation')>.exit_code
> 
> tests/pfcwd_test.py:292: AssertionError
> ----------------------------- Captured stdout call -----------------------------
> Changed polling interval to 199ms on asic0
> BIG_RED_SWITCH status is enable on asic0
> Changed polling interval to 199ms on asic1
> BIG_RED_SWITCH status is enable on asic1
>           PORT    ACTION    DETECTION TIME    RESTORATION TIME
> --------------  --------  ----------------  ------------------
>      Ethernet0      drop               200                 200
>      Ethernet4      drop               200                 200
>   Ethernet-BP0      drop               200                 200
>   Ethernet-BP4      drop               200                 200
> Ethernet-BP256      drop               200                 200
> Ethernet-BP260      drop               200                 200
> 
> Root privileges are required for this operation
> 
> ____________ TestMultiAsicPfcwdShow.test_pfcwd_start_actions_masic _____________
> 
> self = <tests.pfcwd_test.TestMultiAsicPfcwdShow object at 0x7f1221eb7400>
> 
>     def test_pfcwd_start_actions_masic(self):
>         # pfcwd start --action drop --restoration-time 200 Ethernet0 200
>         import pfcwd.main as pfcwd
>         runner = CliRunner()
>         db = Db()
>         # get initial config
>         result = runner.invoke(
>             pfcwd.cli.commands["show"].commands["config"],
>             obj=db
>         )
>         print(result.output)
>         assert result.output == show_pfc_config_all
>     
>         result = runner.invoke(
>             pfcwd.cli.commands["start"],
>             [
>                 "--action", "drop", "--restoration-time", "301",
>                 "all", "302"
>             ],
>             obj=db
>         )
>         print(result.output)
> >       assert result.exit_code == 0
> E       AssertionError: assert 1 == 0
> E        +  where 1 = <Result SystemExit('Root privileges are required for this operation')>.exit_code
> 
> tests/pfcwd_test.py:325: AssertionError
> ```

can you run with "sudo" option. 

What command are you using?

**Comment by stephenxs** [comment]
I just tried building the sonic utilities python wheel package. Something like “make target/python-wheels/sonic-utilities-xx”(the full path name of the package).
The unit test script will be executed during compiling.

**Comment by stephenxs** [comment]
I just tried building the sonic utilities python wheel package. Something like “make target/python-wheels/sonic-utilities-xx”(the full path name of the package).
The unit test script will be executed during compiling.
We don’t use “sudo” when compiling the code.

**Comment by smaheshm** [comment]
"sudo" is required to run the 'pfcwd config' tests. Id suggest to add a sudo check in the test script and skip if not root.

```
if os.geteuid() != 0:
     pytest.skip("need sudo access")
```

**Comment by stephenxs** [comment]
I think to skip it if it's not prilivege user is a work around for now. But I think it's better to support testing it without "sudo".
Probably you can refer how unit test is execucted on "config" command?

**Comment by smaheshm** [comment]
We should not add any checks to skip 'root' privilege in CLI code, since any user can simulate a test environment on a production switch and bypass the 'root' privilege to make config changes. I'm open for any other suggestions.

**Comment by stephenxs** [comment]
sounds reasonable...

**Comment by stephenxs** [comment]
Is it possible to use something like `fakeroot` during the compiling in order to simulate a root user?

**Comment by smaheshm** [comment]
> Is it possible to use something like `fakeroot` during the compiling in order to simulate a root user?

FYI.

https://github.com/Azure/sonic-utilities/pull/1340

**Pull request merged**